modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/ExtensionComponent.java

```diff
@@ -149,16 +149,24 @@ public void setFlowConstruct(FlowConstruct flowConstruct)
 
     private MetadataContext getMetadataContext() throws MetadataResolvingException
     {
-        //TODO MULE-9530: Improve Config retrieval for Metadata resolution
-        if (!StringUtils.isBlank(configurationProviderName) &&
-            muleContext.getRegistry().get(configurationProviderName) instanceof DynamicConfigurationProvider)
+        ConfigurationInstance<Object> configuration;
+        Optional<ConfigurationProvider<Object>> provider = getConfigurationProvider();
+        if (provider.isPresent())
+        {
+            ConfigurationProvider<Object> configurationProvider = provider.get();
+            if (configurationProvider instanceof DynamicConfigurationProvider)
+            {
+                throw new MetadataResolvingException("Configuration used for Metadata fetch cannot be dynamic", FailureCode.INVALID_CONFIGURATION);
+            }
+
+            configuration = configurationProvider.get(getInitialiserEvent(muleContext));
+        }
+        else
         {
-            throw new MetadataResolvingException("Configuration used for Metadata fetch cannot be dynamic", FailureCode.INVALID_CONFIGURATION);
+            configuration = extensionManager.getConfiguration(extensionModel, getInitialiserEvent(muleContext));
```

Comment:
this doesn't look correct. This method will try to create an implicit config if none available. That config might be dynamic. It looks like you should do this if the provider is not present, and move the validation about it not being dynamic below it